src/compare.c

```diff
@@ -192,19 +44,32 @@ SEXP vctrs_compare(SEXP x, SEXP y, SEXP na_equal_) {
   x = PROTECT(vec_normalize_encoding(x));
   y = PROTECT(vec_normalize_encoding(y));
 
-  switch (type) {
-  case vctrs_type_logical:   COMPARE(int, LOGICAL_RO, lgl_compare_scalar);
-  case vctrs_type_integer:   COMPARE(int, INTEGER_RO, int_compare_scalar);
-  case vctrs_type_double:    COMPARE(double, REAL_RO, dbl_compare_scalar);
-  case vctrs_type_character: COMPARE(SEXP, STRING_PTR_RO, chr_compare_scalar);
-  case vctrs_type_dataframe: {
+  if (type == vctrs_type_dataframe) {
     SEXP out = df_compare(x, y, na_equal, size);
     UNPROTECT(2);
     return out;
   }
-  case vctrs_type_scalar:    Rf_errorcall(R_NilValue, "Can't compare scalars with `vctrs_compare()`");
-  case vctrs_type_list:      Rf_errorcall(R_NilValue, "Can't compare lists with `vctrs_compare()`");
-  default:                   Rf_error("Unimplemented type in `vctrs_compare()`");
+
+  if (na_equal) {
+    switch (type) {
+    case vctrs_type_logical:   COMPARE(int, LOGICAL_RO, lgl_compare_na_equal);
+    case vctrs_type_integer:   COMPARE(int, INTEGER_RO, int_compare_na_equal);
+    case vctrs_type_double:    COMPARE(double, REAL_RO, dbl_compare_na_equal);
+    case vctrs_type_character: COMPARE(SEXP, STRING_PTR_RO, chr_compare_na_equal);
+    case vctrs_type_scalar:    Rf_errorcall(R_NilValue, "Can't compare scalars with `vctrs_compare()`");
+    case vctrs_type_list:      Rf_errorcall(R_NilValue, "Can't compare lists with `vctrs_compare()`");
+    default:                   Rf_error("Unimplemented type in `vctrs_compare()`");
+    }
+  } else {
+    switch (type) {
+    case vctrs_type_logical:   COMPARE(int, LOGICAL_RO, lgl_compare_na_propagate);
+    case vctrs_type_integer:   COMPARE(int, INTEGER_RO, int_compare_na_propagate);
+    case vctrs_type_double:    COMPARE(double, REAL_RO, dbl_compare_na_propagate);
+    case vctrs_type_character: COMPARE(SEXP, STRING_PTR_RO, chr_compare_na_propagate);
+    case vctrs_type_scalar:    Rf_errorcall(R_NilValue, "Can't compare scalars with `vctrs_compare()`");
+    case vctrs_type_list:      Rf_errorcall(R_NilValue, "Can't compare lists with `vctrs_compare()`");
+    default:                   Rf_error("Unimplemented type in `vctrs_compare()`");
```

Comment:
Could take the opportunity to switch to `r_abort()` and `stop_unimplemented_vctrs_type()`.